src/cage/RwaCageSettlement.t.sol

```diff
@@ -0,0 +1,481 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity 0.6.12;
+
+import {Test} from "forge-std/Test.sol";
+import {DSToken} from "ds-token/token.sol";
+import {DSMath} from "ds-math/math.sol";
+import {Vat} from "dss/vat.sol";
+import {RwaCageSettlement} from "./RwaCageSettlement.sol";
+
+contract RwaCageSettlementTest is Test {
+    uint256 internal constant WAD = 10**18;
+
+    RwaCageSettlement internal settlement;
+    TestCoin internal gem;
+    TestCoin internal coin;
+    address internal sender = address(0x1337);
+
+    // Fuzz test parameters
+    uint8 internal coinDecimals;
+    uint8 internal gemDecimals;
+
+    // Regular test parameters
+    uint8 internal constant DEFAULT_COIN_DECIMALS = 6;
+    uint8 internal constant DEFAULT_GEM_DECIMALS = 18;
+    uint256 internal constant DEFAULT_GEM_PRICE = 100_000_000 * WAD;
+
+    // Using regular setUp makes fuzz tests fail, so we need a different name.
+    // We also need to invoke this function manually in every test.
+    function _setUp() internal {
+        _parametricSetUp(DEFAULT_COIN_DECIMALS, DEFAULT_GEM_DECIMALS, DEFAULT_GEM_PRICE);
+    }
+
+    function _parametricSetUp(
+        uint8 _coinDecimals,
+        uint8 _gemDecimals,
+        uint256 _gemPrice
+    ) internal {
+        require(_coinDecimals <= _gemDecimals, "coinDecimals too big");
+        require(_gemDecimals <= 27, "gemDecimals too big");
+        require(_gemPrice >= 1 * WAD && _gemPrice <= 10**15 * WAD, "gemPrice out of bounds");
+
+        coinDecimals = _coinDecimals;
+        gemDecimals = _gemDecimals;
+
+        coin = new TestCoin("USDX", coinDecimals);
+        gem = new TestCoin("RWA00X", gemDecimals);
+        settlement = new RwaCageSettlement(address(coin), address(gem), _gemPrice);
+
+        // Mint the gem into the sender's balance
+        uint256 gemSupply = 1 * 10**uint256(gemDecimals);
+        _mint(gem, sender, gemSupply);
+
+        // Mint enough coins into the settlements's balance
+        uint256 requiredCoinAmt = settlement.totalRequiredDeposit();
+        _mint(coin, address(settlement), requiredCoinAmt);
+
+        // Approve the settlement to spend RWA tokens on sender's behalf
+        _approve(address(gem), sender, address(settlement), type(uint256).max);
+
+        // Approve the settlement to spend coins on sender's behalf
+        _approve(address(coin), sender, address(settlement), type(uint256).max);
+    }
+
+    // Asserts against gem and coin balance changes of the sender and the settlement
+    // contract, before and after calling `redeem` with an arbitrary value.
+    function testRedeemFuzz(
+        uint8 _coinDecimals,
+        uint8 _gemDecimals,
+        uint256 _gemPrice,
+        uint256 _redeemGems
+    ) public {
+        // [2,27] decimals
+        _gemDecimals = 2 + (_gemDecimals % (27 - 2 + 1));
```

Comment:
We can use `bound` cheatcode from foundry here
https://book.getfoundry.sh/reference/forge-std/bound

Reply:
Oh, that's a new one... I'll give it a try... thank you (: 👍🏻